pkg/mutation/system.go

```diff
@@ -0,0 +1,165 @@
+package mutation
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/mutator"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/schema"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// System keeps the list of mutations and
+// provides an interface to apply mutations.
+type System struct {
+	schemaDB schema.DB
+	mutators []mutator.Mutator
+	sync.RWMutex
+}
+
+// NewSystem initializes an empty mutation system
+func NewSystem() *System {
+	return &System{
+		mutators: make([]mutator.Mutator, 0),
+	}
+}
+
+// Upsert updates or insert the given object, and returns
+// an error in case of conflicts
+func (s *System) Upsert(m mutator.Mutator) error {
+	s.Lock()
+	defer s.Unlock()
+
+	// Checking schema consistence only if the mutator has schema
+	if withSchema, ok := m.(mutator.WithSchema); ok {
+		err := s.schemaDB.Upsert(withSchema)
+		if err != nil {
+			return errors.Wrapf(err, "Schema upsert failed")
+		}
+	}
+
+	i := sort.Search(len(s.mutators), func(i int) bool {
+		return greaterOrEqual(s.mutators[i], m)
+	})
+
+	if i == len(s.mutators) { // Adding to the bottom of the list
+		s.mutators = append(s.mutators, m)
+		return nil
+	}
+
+	found := equal(s.mutators[i], m)
+
+	if found {
+		if !m.HasDiff(s.mutators[i]) { // no relevant changes, skipping
+			return nil
+		}
+		s.mutators[i] = m
+		return nil
+	}
+
+	s.mutators = append(s.mutators, nil)
+	copy(s.mutators[i+1:], s.mutators[i:])
+	s.mutators[i] = m
+	return nil
+}
+
+// Mutate applies the mutation in place to the given object
+func (s *System) Mutate(obj *unstructured.Unstructured, ns *corev1.Namespace) error {
+	s.RLock()
+	defer s.RUnlock()
+
+	maxIterations := len(s.mutators) + 1
+
+	old := &unstructured.Unstructured{}
+
+	for i := 0; i < maxIterations; i++ {
+		for _, m := range s.mutators {
+			if m.Matches(obj, ns) {
+				err := m.Mutate(obj)
+				if err != nil {
+					return errors.Wrapf(err, "Mutation failed for %s %s", obj.GroupVersionKind().Group, obj.GroupVersionKind().Kind)
+				}
+			}
+		}
+		if cmp.Equal(old, obj) {
+			return nil
+		}
+		old = obj.DeepCopy()
+	}
+
+	return fmt.Errorf("Mutation not converging for %s %s", obj.GroupVersionKind().Group, obj.GroupVersionKind().Kind)
+}
+
+// Remove removes the mutator from the mutation system
+func (s *System) Remove(m mutator.Mutator) error {
+	s.Lock()
+	defer s.Unlock()
+
+	// Checking schema consistence only if the mutator has schema
+	if withSchema, ok := m.(mutator.WithSchema); ok {
```

Comment:
I wonder -- if db.Remove() returns no-error if an object doesn't exist, is this check necessary?

Reply:
You are right, removing it.